In [74]:
#libraries used
import pandas as pd

In [7]:
#load data 
data = pd.read_csv("../data/raw/Citywide_Survey.csv")

In [8]:
#disply data
data

hh_id     weight        cms_zone  survey_mode    person_id  \
0     191546322    0.00000  Inner Brooklyn            1  19154632203   
1     191546322    0.00000  Inner Brooklyn            1  19154632202   
2     191546322  212.12729  Inner Brooklyn            1  19154632201   
3      19241657    0.00000   Middle Queens            1   1924165703   
4      19241657    0.00000   Middle Queens            1   1924165702   
...         ...        ...             ...          ...          ...   
8281   19376768    0.00000   Staten Island            2   1937676802   
8282   19104727   72.75091   Staten Island            2   1910472701   
8283   19104727    0.00000   Staten Island            2   1910472704   
8284   19104727    0.00000   Staten Island            2   1910472703   
8285   19104727    0.00000   Staten Island            2   1910472702   

      person_num  is_participant  num_days  num_days_complete  num_trips  ...  \
0              3               0       NaN                NaN        NaN  ...   
1              2               0       NaN                NaN        NaN  ...   
2              1               1       7.0                7.0       23.0  ...   
3              3               0       NaN                NaN        NaN  ...   
4              2               0       NaN                NaN        NaN  ...   
...          ...             ...       ...                ...        ...  ...   
8281           2               0       NaN                NaN        NaN  ...   
8282           1               1       1.0                1.0        2.0  ...   
8283           4               0       NaN                NaN        NaN  ...   
8284           3               0       NaN                NaN        NaN  ...   
8285           2               0       NaN                NaN        NaN  ...   

      race_native_hawaiian  race_white  race_other  disability_hearing  \
0                      995         995         995                 995   
1                      995         995         995                 995   
2                        1           0           0                   0   
3                      995         995         995                 995   
4                      995         995         995                 995   
...                    ...         ...         ...                 ...   
8281                   995         995         995                 995   
8282                     0           1           0                   0   
8283                   995         995         995                 995   
8284                   995         995         995                 995   
8285                   995         995         995                 995   

      disability_seeing  disability_walking  disability_mobility  \
0                   995                 995                  995   
1                   995                 995                  995   
2                     0                   0                    0   
3                   995                 995                  995   
4                   995                 995                  995   
...                 ...                 ...                  ...   
8281                995                 995                  995   
8282                  0                   0                    0   
8283                995                 995                  995   
8284                995                 995                  995   
8285                995                 995                  995   

     disability_dressing disability_concentrating  disability_none  
0                    995                      995              995  
1                    995                      995              995  
2                      0                        0                1  
3                    995                      995              995  
4                    995                      995              995  
...                  ...                      ...   

### Grouping by Houshold Id

In [69]:
household_grouped = data.groupby("hh_id").agg(set)

In [70]:
# display result
household_grouped

weight              cms_zone survey_mode  \
hh_id                                                            
19000235   {0.0, 1822.20735}      {Northern Bronx}         {1}   
19000474     {0.0, 116.5292}      {Southern Bronx}         {2}   
19001583          {99.99808}      {Northern Bronx}         {1}   
19001694     {0.0, 75.47702}       {Staten Island}         {1}   
19001990    {0.0, 261.31593}  {Northern Manhattan}         {1}   
...                      ...                   ...         ...   
199599840         {306.9369}      {Manhattan Core}         {2}   
199703066        {709.00528}      {Manhattan Core}         {1}   
199794849   {0.0, 404.46156}      {Inner Brooklyn}         {1}   
199848951        {458.66968}  {Northern Manhattan}         {1}   
199989189   {0.0, 227.53115}      {Inner Brooklyn}         {1}   

                                                   person_id       person_num  \
hh_id                                                                           
19000235   {1900023501, 1900023502, 1900023503, 190002350...  {1, 2, 3, 4, 5}   
19000474                {1900047401, 1900047402, 1900047403}        {1, 2, 3}   
19001583                                        {1900158301}              {1}   
19001694                            {1900169401, 1900169402}           {1, 2}   
19001990                            {1900199001, 1900199002}           {1, 2}   
...                                                      ...              ...   
199599840                                      {19959984001}              {1}   
199703066                                      {19970306601}              {1}   
199794849  {19979484901, 19979484902, 19979484903, 199794...  {1, 2, 3, 4, 5}   
199848951                                      {19984895101}              {1}   
199989189                         {19998918901, 19998918902}           {1, 2}   

          is_participant                   num_days  \
hh_id                                                 
19000235          {0, 1}  {nan, nan, nan, nan, 7.0}   
19000474          {0, 1}            {nan, 1.0, nan}   
19001583             {1}                      {7.0}   
19001694          {0, 1}                 {nan, 7.0}   
19001990          {0, 1}                 {nan, 7.0}   
...                  ...                        ...   
199599840            {1}                      {1.0}   
199703066            {1}                      {7.0}   
199794849         {0, 1}  {nan, nan, nan, nan, 7.0}   
199848951            {1}                      {7.0}   
199989189         {0, 1}                 {nan, 7.0}   

                   num_days_complete                   num_trips  \
hh_id                                                              
19000235   {nan, nan, 2.0, nan, nan}  {nan, nan, nan, nan, 23.0}   
19000474             {nan, 1.0, nan}             {0.0, nan, nan}   
19001583                       {7.0}                      {25.0}   
19001694                  {nan, 7.0}                 {nan, 19.0}   
19001990                  {nan, 1.0}                 {nan, 28.0}   
...                              ...                         ...   
199599840                      {1.0}                       {3.0}   
199703066                      {6.0}                      {53.0}   
199794849  {nan, nan, nan, nan, 7.0}  {nan, nan, nan, nan, 57.0}   
199848951                      {3.0}                      {11.0}   
199989189                 {nan, 7.0}                 {nan, 66.0}   

                       num_walk_trips  ... race_native_hawaiian race_white  \
hh_id                                  ...                                   
19000235    {nan, nan, 2.0, nan, nan}  ...             {0, 995}   {0, 995}   
19000474              {0.0, nan, nan}  ...             {0, 995}   {0, 995}   
19001583                       {10.0}  ...                  {0}        {0}   
19001694                   {nan, 0.0}  ...             {0, 995}   {1, 995}   
19001990                   {nan, 0.0}  .

### There are many features (columns) that have values just for one member of the household and the other members have values ( 0 or nan or 995 etc) which actually must be the same value as the member with the  real value. 

In [116]:
# grouped data without nulls
without_null = data.groupby('hh_id').agg(lambda cell: set(cell.dropna()))

In [115]:
#display result
without_null

weight              cms_zone survey_mode  \
hh_id                                                            
19000235   {0.0, 1822.20735}      {Northern Bronx}         {1}   
19000474     {0.0, 116.5292}      {Southern Bronx}         {2}   
19001583          {99.99808}      {Northern Bronx}         {1}   
19001694     {0.0, 75.47702}       {Staten Island}         {1}   
19001990    {0.0, 261.31593}  {Northern Manhattan}         {1}   
...                      ...                   ...         ...   
199599840         {306.9369}      {Manhattan Core}         {2}   
199703066        {709.00528}      {Manhattan Core}         {1}   
199794849   {0.0, 404.46156}      {Inner Brooklyn}         {1}   
199848951        {458.66968}  {Northern Manhattan}         {1}   
199989189   {0.0, 227.53115}      {Inner Brooklyn}         {1}   

                                                   person_id       person_num  \
hh_id                                                                           
19000235   {1900023501, 1900023502, 1900023503, 190002350...  {1, 2, 3, 4, 5}   
19000474                {1900047401, 1900047402, 1900047403}        {1, 2, 3}   
19001583                                        {1900158301}              {1}   
19001694                            {1900169401, 1900169402}           {1, 2}   
19001990                            {1900199001, 1900199002}           {1, 2}   
...                                                      ...              ...   
199599840                                      {19959984001}              {1}   
199703066                                      {19970306601}              {1}   
199794849  {19979484901, 19979484902, 19979484903, 199794...  {1, 2, 3, 4, 5}   
199848951                                      {19984895101}              {1}   
199989189                         {19998918901, 19998918902}           {1, 2}   

          is_participant num_days num_days_complete num_trips num_walk_trips  \
hh_id                                                                          
19000235          {0, 1}    {7.0}             {2.0}    {23.0}          {2.0}   
19000474          {0, 1}    {1.0}             {1.0}     {0.0}          {0.0}   
19001583             {1}    {7.0}             {7.0}    {25.0}         {10.0}   
19001694          {0, 1}    {7.0}             {7.0}    {19.0}          {0.0}   
19001990          {0, 1}    {7.0}             {1.0}    {28.0}          {0.0}   
...                  ...      ...               ...       ...            ...   
199599840            {1}    {1.0}             {1.0}     {3.0}          {3.0}   
199703066            {1}    {7.0}             {6.0}    {53.0}         {23.0}   
199794849         {0, 1}    {7.0}             {7.0}    {57.0}         {39.0}   
199848951            {1}    {7.0}             {3.0}    {11.0}          {5.0}   
199989189         {0, 1}    {7.0}             {7.0}    {66.0}         {38.0}   

           ... race_native_hawaiian race_white race_other disability_hearing  \
hh_id      ...                                                                 
19000235   ...             {0, 995}   {0, 995}   {0, 995}           {0, 995}   
19000474   ...             {0, 995}   {0, 995}   {0, 995}           {0, 995}   
19001583   ...                  {0}        {0}        {0}                {0}   
19001694   ...             {0, 995}   {1, 995}   {0, 995}           {0, 995}   
19001990   ...             {1, 995}   {0, 995}   {0, 995}           {0, 995}   
...        ...                  ...        ...        ...                ...   
199599840  ...                  {0}        {0}        {0}                {0}   
199703066  ...                  {0}        {1}        {0}                {0}   
199794849  ...             {0, 995}   {1, 995}   {0, 995}           {0, 995}   
199848951  ...                  {0}        {0}        {0}                {0}   
199989189  ...             {0, 995}   {1, 995}   {0, 995}           {0, 995}   

          disability_seeing

### The result of grouping by hh_id ==> the number of the households equal 3346

### Groubing by person_id 

In [46]:
person_grouped = data.groupby("person_id").agg(list)

In [47]:
# disply result
person_grouped

hh_id        weight              cms_zone survey_mode  \
person_id                                                                  
1900023501    [19000235]  [1822.20735]      [Northern Bronx]         [1]   
1900023502    [19000235]         [0.0]      [Northern Bronx]         [1]   
1900023503    [19000235]         [0.0]      [Northern Bronx]         [1]   
1900023504    [19000235]         [0.0]      [Northern Bronx]         [1]   
1900023505    [19000235]         [0.0]      [Northern Bronx]         [1]   
...                  ...           ...                   ...         ...   
19979484904  [199794849]         [0.0]      [Inner Brooklyn]         [1]   
19979484905  [199794849]         [0.0]      [Inner Brooklyn]         [1]   
19984895101  [199848951]   [458.66968]  [Northern Manhattan]         [1]   
19998918901  [199989189]   [227.53115]      [Inner Brooklyn]         [1]   
19998918902  [199989189]         [0.0]      [Inner Brooklyn]         [1]   

            person_num is_participant num_days num_days_complete num_trips  \
person_id                                                                    
1900023501         [1]            [1]    [7.0]             [2.0]    [23.0]   
1900023502         [2]            [0]    [nan]             [nan]     [nan]   
1900023503         [3]            [0]    [nan]             [nan]     [nan]   
1900023504         [4]            [0]    [nan]             [nan]     [nan]   
1900023505         [5]            [0]    [nan]             [nan]     [nan]   
...                ...            ...      ...               ...       ...   
19979484904        [4]            [0]    [nan]             [nan]     [nan]   
19979484905        [5]            [0]    [nan]             [nan]     [nan]   
19984895101        [1]            [1]    [7.0]             [3.0]    [11.0]   
19998918901        [1]            [1]    [7.0]             [7.0]    [66.0]   
19998918902        [2]            [0]    [nan]             [nan]     [nan]   

            num_walk_trips  ... race_native_hawaiian race_white race_other  \
person_id                   ...                                              
1900023501           [2.0]  ...                  [0]        [0]        [0]   
1900023502           [nan]  ...                [995]      [995]      [995]   
1900023503           [nan]  ...                [995]      [995]      [995]   
1900023504           [nan]  ...                [995]      [995]      [995]   
1900023505           [nan]  ...                [995]      [995]      [995]   
...                    ...  ...                  ...        ...        ...   
19979484904          [nan]  ...                [995]      [995]      [995]   
19979484905          [nan]  ...                [995]      [995]      [995]   
19984895101          [5.0]  ...                  [0]        [0]        [0]   
19998918901         [38.0]  ...                  [0]        [1]        [0]   
19998918902          [nan]  ...                [995]      [995]      [995]   

            disability_hearing disability_seeing disability_walking  \
person_id                                                             
1900023501                 [0]               [0]                [0]   
1900023502               [995]             [995]              [995]   
1900023503               [995]             [995]              [995]   
1900023504               [995]             [995]              [995]   
1900023505               [995]             [995]              [995]   
...                        ...               ...                ...   
19979484904              [995]             [995]              [995]   
19979484905              [995]             [995]              [995]   
19984895101                [0]               [0]                [0]   
19998918901                [0]               [0]                [0]   
19998918902              [995]             [995]              [995]   

            disability_mobility disability_dressing disability_concentrating

### The result of groubing by person id ==>  equal 8286 same size of data set (unique)